backend/utils/misc.py

```diff
@@ -135,26 +135,55 @@ def apply_model_system_prompt_to_body(params: dict, form_data: dict, user) -> di
 
 
 # inplace function: form_data is modified
-def apply_model_params_to_body(params: dict, form_data: dict) -> dict:
+def apply_model_params_to_body(
+    params: dict, form_data: dict, mappings: dict[str, Callable]
+) -> dict:
     if not params:
         return form_data
 
-    mappings = {
-        "temperature": float,
-        "top_p": int,
-        "max_tokens": int,
-        "frequency_penalty": int,
-        "seed": lambda x: x,
-        "stop": lambda x: [bytes(s, "utf-8").decode("unicode_escape") for s in x],
-    }
-
     for key, cast_func in mappings.items():
         if (value := params.get(key)) is not None:
             form_data[key] = cast_func(value)
 
     return form_data
 
 
+OPENAI_MAPPINGS = {
+    "temperature": float,
+    "top_p": int,
+    "max_tokens": int,
+    "frequency_penalty": int,
+    "seed": lambda x: x,
+    "stop": lambda x: [bytes(s, "utf-8").decode("unicode_escape") for s in x],
+}
+
+
+# inplace function: form_data is modified
+def apply_model_params_to_body_openai(params: dict, form_data: dict) -> dict:
+    return apply_model_params_to_body(params, form_data, OPENAI_MAPPINGS)
+
+
+def apply_model_params_to_body_ollama(params: dict, form_data: dict) -> dict:
```

Comment:
`repeat_penalty` is not being properly applied it seems.

Reply:
Good catch! Same with `max_tokens` tokens it seems like.

It turns out the panel on the right changed `payload` rather than `model_info`, so I wasn't testing what I thought I was testing.

On that note, how does one change `model_info` for an Ollama model? Is that in the Ollama modelfile? If so, shouldn't the per-chat controls take priority over that rather than being overwritten by `model_info` like the way it is now?

Reply:
It's obtained from the values set from the model workspace.